File: backup/Search/c++/18_고정점찾기.cpp
// 이것이 취업을 위한 코딩테스트다
// 1

#include <iostream>
#include <vector>

using namespace std;

int N;
vector<int> arr;

int binary_search(vector<int> arr, int start, int end) {
	if (start > end)
		return -1;

	int mid = (start + end) / 2;
	if (mid == arr[mid])
		return mid;
	else if (arr[mid] >= mid)
		return binary_search(arr, start, mid - 1);
	else
		return binary_search(arr, mid + 1, end);
}

int main() {
	cin >> N;
	for (int i = 0; i < N; i++) {
		int x;
		cin >> x;
		arr.push_back(x);
	}

	cout << binary_search(arr, 0, N - 1);
}

File: backup/Search/python/18_고정점찾기.py
# 이것이 취업을 위한 코딩테스트다
# 1

def binary_search(arr, start, end):
    if start > end:
        return -1
    
    mid = (start + end) // 2
    if arr[mid] == mid:
        return mid
    elif arr[mid] > mid:
        return binary_search(arr, start, mid - 1)
    else:
        return binary_search(arr, mid + 1, end)

N = int(input())
arr = list(map(int, input().split()))

print(binary_search(arr, 0, N - 1))
